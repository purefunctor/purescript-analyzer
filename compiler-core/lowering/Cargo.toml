[package]
name = "lowering"
version = "0.1.0"
edition = "2021"

[dependencies]
hashbrown = "0.15.2"
indexing = { version = "0.1.0", path = "../indexing" }
indexmap = "2.7.1"
itertools = "0.14.0"
la-arena = "0.3.1"
paste = "1.0.15"
rowan = "0.16.1"
rustc-hash = "2.1.1"
smol_str = "0.3.2"
syntax = { version = "0.1.0", path = "../syntax" }

[dev-dependencies]
insta = "1.42.1"
lexing = { path = "../lexing" }
parsing = { path = "../parsing" }
test_each_file = "0.3.4"
