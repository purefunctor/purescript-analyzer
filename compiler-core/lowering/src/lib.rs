mod algorithm;

pub mod intermediate;
pub mod scope;
pub mod source;

pub use intermediate::*;
pub use scope::*;
pub use source::*;

use indexing::FullIndexedModule;
use syntax::cst;

#[derive(Debug, PartialEq, Eq)]
pub struct FullLoweredModule {
    pub intermediate: Intermediate,
    pub source: LoweringSource,
    pub graph: LoweringGraph,
    pub graph_info: LoweringGraphInfo,
}

pub fn lower_module(module: &cst::Module, indexed: &FullIndexedModule) -> FullLoweredModule {
    let algorithm::State { intermediate, source, graph, graph_info, .. } =
        algorithm::lower_module(module, indexed);
    FullLoweredModule { intermediate, source, graph, graph_info }
}
