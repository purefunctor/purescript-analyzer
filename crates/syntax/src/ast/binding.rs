use rowan::ast::{support, AstChildren, AstNode};

use super::{Binder, Expression, LayoutList, LetBinding};

_create_ast_v!(Binding, UnconditionalBinding(UnconditionalBinding), GuardedBinding(GuardedBinding));

_create_ast!(PatternGuard, PatternGuardList, GuardedExpression, WhereExpression);

impl UnconditionalBinding {
    pub fn where_expression(&self) -> Option<WhereExpression> {
        WhereExpression::cast(self.node.first_child()?)
    }
}

impl GuardedBinding {
    pub fn guarded_expressions(&self) -> AstChildren<GuardedExpression> {
        support::children(&self.node)
    }
}

impl PatternGuard {
    pub fn binder(&self) -> Option<Binder> {
        Binder::cast(self.node.first_child()?)
    }

    pub fn expression(&self) -> Option<Expression> {
        Expression::cast(self.node.last_child()?)
    }
}

impl PatternGuardList {
    pub fn children(&self) -> AstChildren<PatternGuard> {
        support::children(&self.node)
    }
}

impl GuardedExpression {
    pub fn pattern_guard_list(&self) -> Option<PatternGuardList> {
        PatternGuardList::cast(self.node.first_child()?)
    }

    pub fn where_expression(&self) -> Option<WhereExpression> {
        WhereExpression::cast(self.node.last_child()?)
    }
}

impl WhereExpression {
    pub fn expression(&self) -> Option<Expression> {
        Expression::cast(self.node.first_child()?)
    }

    pub fn let_bindings(&self) -> Option<LayoutList<LetBinding>> {
        LayoutList::cast(self.node.last_child()?)
    }
}
