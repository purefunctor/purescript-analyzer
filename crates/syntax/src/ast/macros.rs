macro_rules! _create_ast {
    ($($kind:ident),+) => {
        $(
            #[derive(Debug, Clone, PartialEq, Eq, Hash)]
            pub struct $kind {
                node: crate::SyntaxNode,
            }

            impl rowan::ast::AstNode for $kind {
                type Language = crate::PureScript;

                fn can_cast(kind: crate::SyntaxKind) -> bool
                where
                    Self: Sized,
                {
                    matches!(kind, crate::SyntaxKind::$kind)
                }

                fn cast(node: crate::SyntaxNode) -> Option<Self>
                where
                    Self: Sized,
                {
                    if Self::can_cast(node.kind()) {
                        Some(Self { node })
                    } else {
                        None
                    }
                }

                fn syntax(&self) -> &crate::SyntaxNode {
                    &self.node
                }
            }
        )+
    };
}

macro_rules! _create_ast_t {
    ($($kind:ident),+) => {
        $(
            #[derive(Debug, Clone, PartialEq, Eq, Hash)]
            pub struct $kind<T> {
                node: crate::SyntaxNode,
                _kind: std::marker::PhantomData<fn() -> T>
            }

            impl<T> rowan::ast::AstNode for $kind<T> {
                type Language = crate::PureScript;

                fn can_cast(kind: crate::SyntaxKind) -> bool
                where
                    Self: Sized,
                {
                    matches!(kind, crate::SyntaxKind::$kind)
                }

                fn cast(node: crate::SyntaxNode) -> Option<Self>
                where
                    Self: Sized,
                {
                    if Self::can_cast(node.kind()) {
                        Some(Self { node, _kind: std::marker::PhantomData::default() })
                    } else {
                        None
                    }
                }

                fn syntax(&self) -> &crate::SyntaxNode {
                    &self.node
                }
            }
        )+
    };
}

macro_rules! _create_ast_t2 {
    ($($kind:ident),+) => {
        $(
            #[derive(Debug, Clone, PartialEq, Eq, Hash)]
            pub struct $kind<T, U> {
                node: crate::SyntaxNode,
                _kind_t: std::marker::PhantomData<fn() -> T>,
                _kind_u: std::marker::PhantomData<fn() -> U>,
            }

            impl<T, U> rowan::ast::AstNode for $kind<T, U> {
                type Language = crate::PureScript;

                fn can_cast(kind: crate::SyntaxKind) -> bool
                where
                    Self: Sized,
                {
                    matches!(kind, crate::SyntaxKind::$kind)
                }

                fn cast(node: crate::SyntaxNode) -> Option<Self>
                where
                    Self: Sized,
                {
                    if Self::can_cast(node.kind()) {
                        Some(Self { node, _kind_t: std::marker::PhantomData::default(), _kind_u: std::marker::PhantomData::default() })
                    } else {
                        None
                    }
                }

                fn syntax(&self) -> &crate::SyntaxNode {
                    &self.node
                }
            }
        )+
    };
}

macro_rules! _create_ast_v {
    ($kind:ident, $key_0:ident($value_0:ident)$(,$key:ident($value:ident))*) => {
        #[derive(Debug, Clone, PartialEq, Eq, Hash)]
        pub enum $kind {
            $key_0($value_0),
            $(
                $key($value),
            )*
        }

        _create_ast!($value_0);
        $(
            _create_ast!($value);
        )*

        impl rowan::ast::AstNode for $kind {
            type Language = crate::PureScript;

            fn can_cast(kind: crate::SyntaxKind) -> bool
            where
                Self: Sized,
            {
                $value_0::can_cast(kind) $(|| $value::can_cast(kind))*
            }

            fn cast(node: crate::SyntaxNode) -> Option<Self>
            where
                Self: Sized,
            {
                if $value_0::can_cast(node.kind()) {
                    Some($kind::$key_0($value_0::cast(node)?))
                } $(else if $value::can_cast(node.kind()) {
                    Some($kind::$key($value::cast(node)?))
                })* else {
                    None
                }
            }

            fn syntax(&self) -> &crate::SyntaxNode {
                match self {
                    $kind::$key_0(t) => t.syntax(),
                    $(
                        $kind::$key(t) => t.syntax(),
                    )*
                }
            }
        }
    };
}

macro_rules! _has_children {
    ($kind_0:ident<$item_0:ident> $(,$kind:ident<$item:ident>)*) => {
        impl $kind_0 {
            pub fn children(&self) -> rowan::ast::AstChildren<$item_0> {
                rowan::ast::support::children(&self.node)
            }
        }

        $(
            impl $kind {
                pub fn children(&self) -> rowan::ast::AstChildren<$item> {
                    rowan::ast::support::children(&self.node)
                }
            }
        )*
    };
}
