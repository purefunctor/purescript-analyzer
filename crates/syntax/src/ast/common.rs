use rowan::ast::{support, AstChildren, AstNode};

use crate::PureScript;

use super::QualifiedName;

_create_ast_t!(Source, Wrapped);

impl<T: AstNode<Language = PureScript>> Source<T> {
    pub fn child(&self) -> Option<T> {
        support::child(&self.node)
    }
}

impl<T: AstNode<Language = PureScript>> Wrapped<T> {
    pub fn child(&self) -> Option<T> {
        support::child(&self.node)
    }
}

_create_ast!(SymbolOperator);
_create_ast_t!(TickOperator);

impl SymbolOperator {
    pub fn name(&self) -> Option<QualifiedName> {
        support::child(&self.node)
    }
}

impl <T: AstNode<Language = PureScript>> TickOperator<T> {
    pub fn term(&self) -> Option<T> {
        support::child(&self.node)
    }
}

_create_ast_t2!(OperatorPair);

impl<T: AstNode<Language = PureScript>, U: AstNode<Language = PureScript>> OperatorPair<T, U> {
    pub fn operator(&self) -> Option<T> {
        T::cast(self.node.first_child()?)
    }

    pub fn term(&self) -> Option<U> {
        U::cast(self.node.last_child()?)
    }
}

_create_ast_t!(ArgumentList, LayoutList);

impl<T: AstNode<Language = PureScript>> ArgumentList<T> {
    pub fn children(&self) -> AstChildren<T> {
        support::children(&self.node)
    }
}

impl<T: AstNode<Language = PureScript>> LayoutList<T> {
    pub fn children(&self) -> AstChildren<T> {
        support::children(&self.node)
    }
}
