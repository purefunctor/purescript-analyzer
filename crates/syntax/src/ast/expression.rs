use rowan::ast::{support, AstChildren, AstNode};

use crate::{PureScript, SyntaxToken};

use super::{
    ArgumentList, BinderList, LayoutList, LetBinding, Name, NameRef, OperatorPair, QualifiedName,
    SymbolOperator, TickOperator, Type,
};

_create_ast_v!(
    Expression,
    AdoExpression(AdoExpression),
    ApplicationExpression(ApplicationExpression),
    CaseExpression(CaseExpression),
    ConstructorExpression(ConstructorExpression),
    DoExpression(DoExpression),
    ExpressionInfixChain(ExpressionInfixChain),
    ExpressionOperatorChain(ExpressionOperatorChain),
    IfThenElseExpression(IfThenElseExpression),
    LambdaExpression(LambdaExpression),
    LetInExpression(LetInExpression),
    LiteralExpression(LiteralExpression),
    OperatorNameExpression(OperatorNameExpression),
    ParenthesizedExpression(ParenthesizedExpression),
    RecordAccessExpression(RecordAccessExpression),
    RecordUpdateExpression(RecordUpdateExpression),
    TypedExpression(TypedExpression),
    VariableExpression(VariableExpression)
);

_create_ast_v!(Argument, TermArgument(TermArgument), TypeArgument(TypeArgument));

_create_ast_v!(RecordItem, RecordField(RecordField), RecordPun(RecordPun));

impl ApplicationExpression {
    pub fn head(&self) -> Option<Expression> {
        Expression::cast(self.node.first_child()?)
    }

    pub fn spine(&self) -> Option<ArgumentList<Argument>> {
        ArgumentList::cast(self.node.last_child()?)
    }
}

impl ConstructorExpression {
    pub fn qualified_name(&self) -> Option<QualifiedName> {
        QualifiedName::cast(self.node.first_child()?)
    }
}

impl TermArgument {
    pub fn expression(&self) -> Option<Expression> {
        Expression::cast(self.node.first_child()?)
    }
}

impl TypeArgument {
    pub fn at(&self) -> Option<SyntaxToken> {
        self.node.first_token()
    }

    pub fn ty(&self) -> Option<Type> {
        Type::cast(self.node.first_child()?)
    }
}

impl ExpressionInfixChain {
    pub fn head(&self) -> Option<Expression> {
        Expression::cast(self.node.first_child()?)
    }

    pub fn tail(&self) -> AstChildren<OperatorPair<TickOperator<Expression>, Expression>> {
        support::children(&self.node)
    }
}

impl ExpressionOperatorChain {
    pub fn head(&self) -> Option<Expression> {
        Expression::cast(self.node.first_child()?)
    }

    pub fn tail(&self) -> AstChildren<OperatorPair<SymbolOperator, Expression>> {
        support::children(&self.node)
    }
}

impl IfThenElseExpression {
    pub fn condition(&self) -> Option<Expression> {
        Expression::cast(self.node.first_child()?)
    }

    pub fn then(&self) -> Option<Expression> {
        Expression::cast(self.node.children().nth(1)?)
    }

    pub fn unless(&self) -> Option<Expression> {
        Expression::cast(self.node.last_child()?)
    }
}

impl LambdaExpression {
    pub fn binders(&self) -> Option<BinderList> {
        BinderList::cast(self.node.first_child()?)
    }

    pub fn body(&self) -> Option<Expression> {
        Expression::cast(self.node.last_child()?)
    }
}

impl LetInExpression {
    pub fn let_bindings(&self) -> Option<LayoutList<LetBinding>> {
        LayoutList::cast(self.node.first_child()?)
    }

    pub fn body(&self) -> Option<Expression> {
        Expression::cast(self.node.last_child()?)
    }
}

impl RecordField {
    pub fn name(&self) -> Option<Name> {
        Name::cast(self.node.first_child()?)
    }

    pub fn value<T>(&self) -> Option<T>
    where
        T: AstNode<Language = PureScript>,
    {
        T::cast(self.node.last_child()?)
    }
}

impl RecordPun {
    pub fn name(&self) -> Option<Name> {
        Name::cast(self.node.first_child()?)
    }

    pub fn name_ref(&self) -> Option<NameRef> {
        NameRef::cast(self.node.first_child()?)
    }
}

impl ParenthesizedExpression {
    pub fn expression(&self) -> Option<Expression> {
        support::child(&self.node)
    }
}

impl VariableExpression {
    pub fn qualified_name(&self) -> Option<QualifiedName> {
        QualifiedName::cast(self.node.first_child()?)
    }
}
