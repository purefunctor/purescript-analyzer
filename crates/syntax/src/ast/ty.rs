//! of the Kingdom

use rowan::ast::{support, AstChildren, AstNode};

use crate::SyntaxToken;

use super::{ArgumentList, Name, NameRef, OperatorPair, QualifiedName, SymbolOperator};

_create_ast!(ForallVariables);

_create_ast_v!(
    TypeVariableBinding,
    TypeVariableKinded(TypeVariableKinded),
    TypeVariableName(TypeVariableName)
);

_create_ast_v!(
    Type,
    ApplicationType(ApplicationType),
    ArrowType(ArrowType),
    ConstrainedType(ConstrainedType),
    ConstructorType(ConstructorType),
    ForallType(ForallType),
    IntegerType(IntegerType),
    KindedType(KindedType),
    OperatorNameType(OperatorNameType),
    ParenthesizedType(ParenthesizedType),
    RecordType(RecordType),
    RowType(RowType),
    StringType(StringType),
    TypeOperatorChain(TypeOperatorChain),
    VariableType(VariableType),
    WildcardType(WildcardType)
);

impl TypeVariableKinded {
    pub fn at_prefix(&self) -> Option<SyntaxToken> {
        self.node.first_token()
    }

    pub fn name(&self) -> Option<Name> {
        Name::cast(self.node.first_child()?)
    }

    pub fn kind(&self) -> Option<Type> {
        Type::cast(self.node.last_child()?)
    }
}

impl TypeVariableName {
    pub fn at_prefix(&self) -> Option<SyntaxToken> {
        self.node.first_token()
    }

    pub fn name(&self) -> Option<Name> {
        Name::cast(self.node.first_child()?)
    }
}

impl ArrowType {
    pub fn argument(&self) -> Option<Type> {
        Type::cast(self.node.first_child()?)
    }

    pub fn result(&self) -> Option<Type> {
        Type::cast(self.node.last_child()?)
    }
}

impl ApplicationType {
    pub fn head(&self) -> Option<Type> {
        Type::cast(self.node.first_child()?)
    }

    pub fn spine(&self) -> Option<ArgumentList<Type>> {
        ArgumentList::cast(self.node.last_child()?)
    }
}

impl ConstrainedType {
    pub fn constraint(&self) -> Option<Type> {
        Type::cast(self.node.first_child()?)
    }

    pub fn constrained(&self) -> Option<Type> {
        Type::cast(self.node.last_child()?)
    }
}

impl ConstructorType {
    pub fn qualified_name(&self) -> Option<QualifiedName> {
        QualifiedName::cast(self.node.first_child()?)
    }
}

impl ForallType {
    pub fn variables(&self) -> Option<ForallVariables> {
        ForallVariables::cast(self.node.first_child()?)
    }

    pub fn inner(&self) -> Option<Type> {
        Type::cast(self.node.last_child()?)
    }
}

_has_children!(ForallVariables<TypeVariableBinding>);

impl ParenthesizedType {
    pub fn ty(&self) -> Option<Type> {
        Type::cast(self.node.first_child()?)
    }
}

impl TypeOperatorChain {
    pub fn head(&self) -> Option<Type> {
        Type::cast(self.node.first_child()?)
    }

    pub fn tail(&self) -> AstChildren<OperatorPair<SymbolOperator, Type>> {
        support::children(&self.node)
    }
}

impl VariableType {
    pub fn name_ref(&self) -> Option<NameRef> {
        NameRef::cast(self.node.first_child()?)
    }
}
