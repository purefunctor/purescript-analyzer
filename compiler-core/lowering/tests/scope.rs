mod shared;

use lowering::{
    ExpressionKind, FullLoweredModule, GraphNode, LetBindingResolution, TermResolution, TypeKind,
    TypeVariableResolution,
};
use rowan::ast::AstNode;
use std::fmt::Write;
use test_each_file::test_each_file;

fn variable_scope_check(content: &str) -> String {
    let (module, _, FullLoweredModule { intermediate, source, graph, .. }) =
        shared::lower_source(content);

    let mut snapshot = String::default();

    writeln!(snapshot, "Expressions:").unwrap();
    writeln!(snapshot).unwrap();
    for (expression_id, _) in intermediate.iter_expression() {
        let Some(ExpressionKind::Variable { resolution, .. }) =
            intermediate.index_expression_kind(expression_id)
        else {
            continue;
        };

        let cst = &source[expression_id];
        let root = module.syntax();

        let node = cst.syntax_node_ptr().to_node(root);
        let text = node.text().to_string();
        let range = node.text_range();

        writeln!(snapshot, "{}@{:?}", text.trim(), range).unwrap();
        if let Some(resolution) = resolution {
            match resolution {
                TermResolution::Global => {
                    writeln!(snapshot, "  resolves to top-level name").unwrap();
                }
                TermResolution::Binder(binder) => {
                    let cst = &source[*binder];
                    let range = cst.syntax_node_ptr().text_range();
                    writeln!(snapshot, "  resolves to binder {range:?}").unwrap();
                }
                TermResolution::Let(LetBindingResolution { signature, equations }) => {
                    if let Some(signature) = signature {
                        let cst = &source[*signature];
                        let range = cst.syntax_node_ptr().text_range();
                        writeln!(snapshot, "  resolves to signature {range:?}").unwrap();
                    }
                    for equation in equations.iter() {
                        let cst = &source[*equation];
                        let range = cst.syntax_node_ptr().text_range();
                        writeln!(snapshot, "  resolves to equation {range:?}").unwrap();
                    }
                }
                TermResolution::AdHoc { .. } => {
                    writeln!(snapshot, "  resolves to top-level name").unwrap();
                }
            }
        } else {
            writeln!(snapshot, "  resolves to nothing").unwrap();
        }
    }

    writeln!(snapshot, "\nTypes:\n").unwrap();

    for (type_id, _) in intermediate.iter_type() {
        let Some(TypeKind::Variable { resolution, .. }) = intermediate.index_type_kind(type_id)
        else {
            continue;
        };
        let cst = &source[type_id];
        let range = cst.syntax_node_ptr().text_range();
        writeln!(snapshot, "{range:?}").unwrap();
        if let Some(resolution) = resolution {
            match resolution {
                TypeVariableResolution::Forall(id) => {
                    let cst = &source[*id];
                    let range = cst.syntax_node_ptr().text_range();
                    writeln!(snapshot, "  resolves to forall {range:?}").unwrap();
                }
                TypeVariableResolution::Implicit { binding, node, id } => {
                    if let GraphNode::Implicit { bindings, .. } = &graph[*node] {
                        let (name, type_ids) =
                            bindings.get_index(*id).expect("invariant violated: invalid index");
                        if *binding {
                            writeln!(snapshot, "  introduces a constraint variable {name:?}")
                                .unwrap();
                        } else {
                            writeln!(snapshot, "  resolves to a constraint variable {name:?}")
                                .unwrap();
                            for &type_id in type_ids {
                                let cst = &source[type_id];
                                let range = cst.syntax_node_ptr().text_range();
                                writeln!(snapshot, "    {range:?}").unwrap();
                            }
                        }
                    } else {
                        writeln!(snapshot, "  did not resolve to constraint variable!").unwrap();
                    }
                }
            }
        } else {
            writeln!(snapshot, "  resolves to nothing").unwrap();
        }
    }

    snapshot
}

test_each_file! { in "./compiler-core/lowering/tests/scope" as variable => |content: &str| {
    let snapshot = variable_scope_check(content);
    insta::assert_snapshot!(snapshot);
}}
