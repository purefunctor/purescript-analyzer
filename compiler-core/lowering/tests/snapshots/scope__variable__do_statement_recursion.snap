---
source: compiler-core/lowering/tests/scope.rs
expression: snapshot
---
Expressions:

y'equation@56..67
  resolves to equation 34..76
a'binder@67..76
  resolves to binder 45..54
pure@93..98
  resolves to top-level name
x'notBinder@98..110
  resolves to top-level name
pure@110..117
  resolves to top-level name
x'notBinder@117..129
  resolves to binder 76..90

Types:
